engine/src/main/java/org/terasology/physics/engine/VoxelWorldSystem.java

```diff
@@ -0,0 +1,145 @@
+import com.badlogic.gdx.physics.bullet.linearmath.btDefaultMotionState;
+import gnu.trove.set.hash.TShortHashSet;
+import org.joml.Matrix4f;
+import org.joml.Vector3f;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.math.JomlUtil;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.physics.StandardCollisionGroup;
+import org.terasology.physics.bullet.BulletPhysics;
+import org.terasology.physics.bullet.shapes.BulletCollisionShape;
+import org.terasology.registry.In;
+import org.terasology.world.OnChangedBlock;
+import org.terasology.world.WorldComponent;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockComponent;
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.ChunkConstants;
+import org.terasology.world.chunks.ChunkProvider;
+import org.terasology.world.chunks.event.BeforeChunkUnload;
+import org.terasology.world.chunks.event.OnChunkLoaded;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.terasology.physics.bullet.BulletPhysics.AABB_SIZE;
+
+@RegisterSystem
+public class VoxelWorldSystem extends BaseComponentSystem {
+
+    @In
+    private PhysicsEngine physics;
+    @In
+    private EntityManager entityManager;
+    @In
+    private WorldProvider worldProvider;
+    @In
+    private ChunkProvider chunkProvider;
+
+    private TShortHashSet registred = new TShortHashSet();
+
+    private btRigidBodyConstructionInfo blockConsInf;
+    private btVoxelShape worldShape;
+    private VoxelCollisionAlgorithmWrapper wrapper;
+    private btRigidBody rigidBody;
+
+    @Override
+    public void initialise() {
+        if (physics instanceof BulletPhysics) {
+            btDiscreteDynamicsWorld discreteDynamicsWorld = ((BulletPhysics) physics).getDiscreteDynamicsWorld();
+
+            wrapper = new VoxelCollisionAlgorithmWrapper(ChunkConstants.SIZE_X, ChunkConstants.SIZE_Y, ChunkConstants.SIZE_Z);
+            worldShape = new btVoxelShape(wrapper, new Vector3f(-AABB_SIZE, -AABB_SIZE, -AABB_SIZE), new Vector3f(AABB_SIZE, AABB_SIZE, AABB_SIZE));
+
+            Matrix4f matrix4f = new Matrix4f();
+            btDefaultMotionState blockMotionState = new btDefaultMotionState(matrix4f);
+
+            blockConsInf = new btRigidBodyConstructionInfo(0, blockMotionState, worldShape, new Vector3f());
+            rigidBody = new btRigidBody(blockConsInf);
+            rigidBody.setCollisionFlags(btCollisionObject.CollisionFlags.CF_STATIC_OBJECT | rigidBody.getCollisionFlags());
+            short mask = (short) (~(StandardCollisionGroup.STATIC.getFlag() | StandardCollisionGroup.LIQUID.getFlag()));
```

Comment:
This and other bits of code in this class is kinda ... wild. And unusual. It might make sense if you're familiar with native physics code but if not I'd find this pretty opaque. More comments at least, perhaps?